packages/@aws-cdk/aws-codepipeline-actions/lib/cloudformation/stackset-action.ts

```diff
@@ -0,0 +1,691 @@
+  ENABLED_WITH_STACK_RETENTION = 'EnabledWithStackRetention'
+}
+
+/**
+ * Base parameters for the StackSet
+ */
+export abstract class StackSetParameters {
+  /**
+   * A list of template parameters for your stack set.
+   *
+   * You must specify all template parameters. Parameters you don't specify will revert
+   * to their `Default` values as specified in the template.
+   *
+   * Specify the names of parameters you want to retain their existing values,
+   * without specifying what those values are, in an array in the second
+   * argument to this function. Use of this feature is discouraged. CDK is for
+   * specifying desired-state infrastructure, and use of this feature makes the
+   * parameter values unmanaged.
+   *
+   * @example
+   *
+   * const parameters = cpactions.StackSetParameters.fromLiteral({
+   *  BucketName: 'my-bucket',
+   *  Asset1: 'true',
+   * });
+   */
+  public static fromLiteral(parameters: Record<string, string>, usePreviousValues?: string[]): StackSetParameters {
+    return new class extends StackSetParameters {
+      public readonly artifactsReferenced: codepipeline.Artifact[] = [];
+
+      render(): string {
+        return [
+          ...Object.entries(parameters).map(([key, value]) =>
+            `ParameterKey=${key},ParameterValue=${value}`),
+          ...(usePreviousValues ?? []).map((key) =>
+            `ParameterKey=${key},UsePreviousValue=true`),
+        ].join(' ');
+      }
+    }();
+  }
+
+  /**
+   * Read the parameters from a JSON file from one of the pipeline's artifacts
+   *
+   * The file needs to contain a list of `{ ParameterKey, ParameterValue, UsePreviousValue }` objects, like
+   * this:
+   *
+   * ```
+   * [
+   *     {
+   *         "ParameterKey": "BucketName",
+   *         "ParameterValue": "my-bucket"
+   *     },
+   *     {
+   *         "ParameterKey": "Asset1",
+   *         "ParameterValue": "true"
+   *     },
+   *     {
+   *         "ParameterKey": "Asset2",
+   *         "UsePreviousValue": true
+   *     }
+   * ]
+   * ```
+   *
+   * You must specify all template parameters. Parameters you don't specify will revert
+   * to their `Default` values as specified in the template.
+   *
+   * For of parameters you want to retain their existing values
+   * without specifying what those values are, set `UsePreviousValue: true`.
+   * Use of this feature is discouraged. CDK is for
+   * specifying desired-state infrastructure, and use of this feature makes the
+   * parameter values unmanaged.
+   */
+  public static fromArtifactPath(artifactPath: codepipeline.ArtifactPath): StackSetParameters {
+    return new class extends StackSetParameters {
+      public artifactsReferenced: codepipeline.Artifact[] = [artifactPath.artifact];
+
+      public render(): string {
+        return artifactPath.location;
+      }
+    }();
+  }
+
+  /**
+   * Artifacts referenced by this parameter set
+   */
+  public abstract readonly artifactsReferenced: codepipeline.Artifact[];
```

Comment:
Same question here (any reason this is not part of a `StackSetParametersConfig` that are returned from `bind()`?).

Also, can we make this `@internal`?

Reply:
See my previous answer.